workflows/magmap.nf

```diff
@@ -1,123 +1,688 @@
 */
 
-// Info required for completion email and summary
-def multiqc_report = []
-
 workflow MAGMAP {
 
+    take:
+    ch_samplesheet  // channel: samplesheet read in from --input
+    ch_versions     // channel: [ path(versions.yml) ]
+
+    main:
+
     ch_versions = Channel.empty()
+    ch_multiqc_files = Channel.empty()
+
+    //
+    // INPUT: if user provides, populate ch_genomeinfo with a table that provides the genomes to filter with sourmash
+    //
+    if ( params.genomeinfo) {
+        Channel
+            .fromPath( params.genomeinfo )
+            .splitCsv( sep: ',', header: true )
+            .set { ch_genomeinfo }
+    }
+
+    CHECK_DUPLICATES( ch_genomeinfo.map{ it.genome_fna }.collect() )
+    //
+    // INPUT: genome info from ncbi
+    //
+    if ( params.ncbi_genome_infos) {
+        Channel
+            .fromPath( params.ncbi_genome_infos )
+            .set { ch_genome_infos }
+    }
+
+    //
+    // INPUT: if user provides, populate ch_indexes
+    //
+    ch_indexes = Channel.empty()
+
+    if ( params.indexes) {
+        Channel
+            .fromPath( params.indexes )
+            .set { ch_indexes }
+    }
+
+    //
+    // INPUT: if user provides, populate ch_metadata
+    //
+    ch_gtdb_metadata = Channel.empty()
+
+    if ( params.gtdb_metadata) {
+        Channel
+            .of(params.gtdb_metadata.split(','))
+            .map { file(it) }
+            .splitCsv( sep: '\t', header: true)
+            .map {
+                [
+                    accno: it.accession - ~/^[A-Z][A-Z]_/,
+                    checkm_completeness: it.checkm_completeness,
+                    checkm_contamination: it.checkm_contamination,
+                    checkm_strain_heterogeneity: it.checkm_strain_heterogeneity,
+                    contig_count: it.contig_count,
+                    genome_size: it.genome_size,
+                    gtdb_genome_representative: it.gtdb_genome_representative,
+                    gtdb_representative: it.gtdb_representative,
+                    gtdb_taxonomy: it.gtdb_taxonomy
+                ]
+            }
+            .set { ch_gtdb_metadata }
+    }
+
+    //
+    // INPUT: CheckM metadata
+    //
+    ch_checkm_metadata = Channel.empty()
+    if ( params.checkm_metadata) {
+        Channel
+            .of(params.checkm_metadata.split(','))
+            .map { file(it) }
+            .splitCsv( sep: '\t', header: true)
+            .map { [ [ it["Bin Id"] ],
+                [
+                    checkm_completeness: it.Completeness,
+                    checkm_contamination: it.Contamination,
+                    contig_count: it["# contigs"],
+                    checkm_strain_heterogeneity: it["Strain heterogeneity"],
+                    genome_size: it["Genome size (bp)"]
+                ] ]
+            }
+            .set { ch_checkm_metadata }
+    }
+
+    //
+    // INPUT: GTDB-Tk metadata
+    //
```

Comment:
This whole section for preparing the GTDBTK/CheckM stuffi s very long, maybe consider a subworkflow?